### ISSUE TITLE
[BUG] CI parallel execution hangs indefinitely on teardown for some retried tests

### ISSUE THREAD

**Issue by CoHuK** [microsoft/playwright#24159]
### System info
- Playwright Version: [v1.35.1]
- Operating System: [Linux, Mac]
- Browser: [Chromium]
- Other info: Random occurance

### Source code

- [ ] I provided exact source code that allows reproducing the issue locally.

or

**Config file**

```js
// playwright.config.ts
import { defineConfig, devices } from '@playwright/test';

import { testPlanFilter } from '../node_modules/allure-playwright/dist/testplan';

/**
 * Read environment variables from file.
 * https://github.com/motdotla/dotenv
 */
// require('dotenv').config();

/**
 * @see https://playwright.dev/docs/test-configuration
 */
export default defineConfig({
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,

  /* Run tests in files in parallel */
  fullyParallel: true,
  grep: testPlanFilter(),

  /* Configure projects for major browsers */
  projects: [
    {
      name: 'Localhost',
      use: {
        adminData: {
          password: '123321',
          url: 'http://127.0.0.1:4001/',
          username: 'pavel@lexense.com',
        },
        baseURL: 'http://127.0.0.1:2001/',
      },
    },
    {
      name: 'Localhost Mobile',
      use: {
        ...devices['Pixel 5'],
        baseURL: 'http://127.0.0.1:2001/',
      },
    },
    {
      name: 'Production',
      use: {
        baseURL: 'https://',
      },
    },
  ],
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: [
    ['html', { open: 'never' }],
    ['list'],
    [
      'allure-playwright',
      {
        detail: false,
        environmentInfo: {
          E2E_NODE_VERSION: process.version,
          E2E_OS: process.platform,
        },
      },
    ],
  ],

  /* Retry on CI only */
  retries: process.env.CI ? 1 : 0,

  testDir: './tests',

  timeout: 70000,

  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    acceptDownloads: true,
    actionTimeout: 10000,
    navigationTimeout: 10000,

    screenshot: 'only-on-failure',
    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: 'on-first-retry',
    video: 'retain-on-failure',
    viewport: { height: 800, width: 1440 },
  },

  /* Parallel tests on CI. */
  workers: process.env.CI ? 6 : 3,
});
```

**Test file (self-contained)**
Reproducible with random tests. The only required conditions are: parallel execution and at least 1 retry

**Steps**
- Run the test in parallel
- Some of the tests should be retried

**Expected**
Execution should stop by itself

**Actual**
The teardown process hangs and gets killed by timeout for some tests, but workers never stop.

Also I can see the next artifacts:
- In CI I'm getting an error about a non-existent screenshot file:
![CleanShot 2023-07-11 at 17 59 53](https://github.com/microsoft/playwright/assets/1978717/1531cd7f-4dd7-47b7-b5c1-7d951d24d114)
- Locally, it just hangs without any issues. But even though the passed test, in this case, is treated as Broken and screenshot+video attached.
![CleanShot 2023-07-11 at 18 05 48](https://github.com/microsoft/playwright/assets/1978717/1df54721-c059-4ace-bbb8-156b7af6db2e)

**Comment by mxschmitt** [comment]
Without a clear reproduction we most likely can't assist much here. This can be a bug inside Playwright that the page/context is not tearing down correctly, but could also just be a timeout exceeds of the context. I'd try increasing the global timeout more and setting the `DEBUG=pw:test` env which will give you logs about which fixture is not tearing down.

**Comment by CoHuK** [comment]
I'll try to find a short reproduction, but it may be tricky. I don't think that increasing a global timeout will do much here as you can see, the test itself takes 14 seconds to execute. And the timeout is 70.

**Comment by denkorz** [comment]
+1

**Comment by marcusNumminen** [comment]
Hi!
In one of our test projects we sometimes got hanged Azure pipelined on all playwright versions above 1.34 and we could really understand why so I did some investigations and wanted to share my findings. I wrote a small reporter plugin to aid my investigations (see below). What I found once I managed to get a “hanging” and compared this to local run where I simulated similar failure (in both cases the same expect failed) is that some if the internal playwright calls was missing in the “hanging” case. We use retries in the test runs and it seems to get stuck when its going to retry the test.

When the pipeline hanged:
```
"---> onStepBegin: [hook][open 'Community Conversations'], After Hooks",
  "---> onStepBegin: [pw:api][open 'Community Conversations'], page.screenshot",
  "---> onStepBegin: [pw:api][open 'Community Conversations'], page.screenshot",
  "---> onStepEnd: [pw:api][open 'Community Conversations'], page.screenshot, ",
  "---> onStepEnd: [pw:api][open 'Community Conversations'], page.screenshot, ",
  "---> onStepBegin: [hook][open 'Community Conversations'], afterEach hook",
  "---> onStepBegin: [pw:api][open 'Community Conversations'], page.close",
  "---> onStepEnd: [pw:api][open 'Community Conversations'], page.close, ",
  "---> onStepEnd: [hook][open 'Community Conversations'], afterEach hook, ",
  "---> onStepEnd: [hook][open 'Community Conversations'], After Hooks, ",
  "--> onTestEnd: \"open 'Community Conversations'\": failed, ==> TestError:
```

When I created a similar error locally:
```
---> onStepBegin: [hook][open 'Community Conversations'], After Hooks
---> onStepBegin: [pw:api][open 'Community Conversations'], page.screenshot
---> onStepBegin: [pw:api][open 'Community Conversations'], page.screenshot
---> onStepEnd: [pw:api][open 'Community Conversations'], page.screenshot,
---> onStepEnd: [pw:api][open 'Community Conversations'], page.screenshot,
---> onStepBegin: [hook][open 'Community Conversations'], afterEach hook
---> onStepBegin: [pw:api][open 'Community Conversations'], page.close
---> onStepEnd: [pw:api][open 'Community Conversations'], page.close,
---> onStepEnd: [hook][open 'Community Conversations'], afterEach hook,
---> onStepBegin: [attach][open 'Community Conversations'], attach "screenshot"
---> onStepEnd: [attach][open 'Community Conversations'], attach "screenshot",
---> onStepBegin: [attach][open 'Community Conversations'], attach "screenshot"
---> onStepEnd: [attach][open 'Community Conversations'], attach "screenshot",
---> onStepBegin: [attach][open 'Community Conversations'], attach "trace"
---> onStepEnd: [attach][open 'Community Conversations'], attach "trace",
---> onStepBegin: [hook][open 'Community Conversations'], afterAll hook
---> onStepBegin: [pw:api][open 'Community Conversations'], locator.isVisible([aria-label="Resource Center"] >> internal:role=button[name="Close"i])
---> onStepEnd: [pw:api][open 'Community Conversations'], locator.isVisible([aria-label="Resource Center"] >> internal:role=button[name="Close"i]),
---> onStepBegin: [expect][open 'Community Conversations'], Unable to close 'Resource Center' dialog
---> onStepEnd: [expect][open 'Community Conversations'], Unable to close 'Resource Center' dialog,
---> onStepBegin: [pw:api][open 'Community Conversations'], page.close
---> onStepEnd: [pw:api][open 'Community Conversations'], page.close,
---> onStepEnd: [hook][open 'Community Conversations'], afterAll hook,
---> onStepEnd: [hook][open 'Community Conversations'], After Hooks,
--> onTestEnd: "open 'Community Conversations'": failed, ==> TestError:
```
We can see that there are missing internal calls in the first one.

I managed to “fix” our problem with the hanging pipeline by changing so that we don’t take traces on every run but only on retries (ie, trace: 'retain-on-failure -> trace: 'on-all-retries' in the conf file). I hope that this could give some insights to the issue. As I mentioned above our tests worked fine until we tested to update to playwright version 1.35 and above.


```
//My debug reporter
import type {FullConfig, FullResult, Reporter, Suite, TestCase, TestResult, TestError, TestStep} from '@playwright/test/reporter';
import { sanitize } from 'sanitize-filename-ts';
import * as path from 'path';
import * as fs from 'fs-extra';


class MyReporter implements Reporter {

  private customOptions: {outputFolder?: string, configDir?: string};
  private logs: {[key: string]: string[]} = {};

  constructor(options = {}) {
    this.customOptions = options;
  }

  onBegin(config: FullConfig, suite: Suite) {
    console.log(`-> onBegin: ${suite.title}`);
  }

  onTestBegin(test: TestCase) {
    const msg = `--> onTestBegin: "${test.title}"`;
    console.log(msg);
    this.logs[test.title] = [msg];
  }

  onStepBegin(test: TestCase, result: TestResult, step: TestStep) {
    const msg = `---> onStepBegin: [${step.category}][${test.title}], ${step.title}`;
    console.log(msg);
    this.logs[test.title].push(msg);
  }

  onStepEnd(test: TestCase, result: TestResult, step: TestStep) {    
    const msg = `---> onStepEnd: [${step.category}][${test.title}], ${step.title}, ${step.error ? this.extractErrorMsg(step.error) : ''}`;
    console.log(msg);
    this.logs[test.title].push(msg);
  }

  onTestEnd(test: TestCase, result: TestResult) {
    const msg = `--> onTestEnd: "${test.title}": ${result.status}, ${result.error ? this.extractErrorMsg(result.error) : ''}`;
    console.log(msg);
    this.logs[test.title].push(msg);
    if (this.customOptions.outputFolder) {
      const fileName = `${sanitize(test.title).replaceAll(' ', '_')}.json`;
      const filePath = path.resolve(this.customOptions.configDir, this.customOptions.outputFolder, fileName);
      fs.outputJSONSync(filePath, this.logs[test.title], {spaces: 2});
    }
  }

  onEnd(result: FullResult) {
    console.log(`-> onEnd: ${result.status}`);
  }

  // onError(error: TestError) {
  //   const msg = `==> onError:\n ${this.extractErrorMsg(error)}`;
  //   console.log(msg);
  // }

  onExit(): Promise<void> {
    console.log(`-> onExit`);
    return new Promise((resolve) => {
      resolve();
    });
  }

  printsToStdio(): boolean {
    return false;
  }

  private extractErrorMsg(error: TestError): string {
    return `==> TestError:\n` +
    `    value: ${error?.value}\n`+
    `    message: ${error?.message}\n`+
    `    location: ${error?.location?.file}, (${error?.location?.line}:${error?.location?.column})\n`+
    `    snippet: \n${error?.snippet}\n`+
    `    stack: \n${error?.stack}\n`;
  }
}

export default MyReporter;
```

**Comment by mxschmitt** [comment]
@marcusNumminen would it be possible to file an issue with a reproduction? We usually don't monitor closed issues. Hanging is definitely not expected and we should look into it.

**Comment by marcusNumminen** [comment]
Hi @mxschmitt.
I tried to reproduce the issue locally but I couldn't and this was the reason why I added my findings here instead of creating a new issue. If I somehow managed to reproduce the issue I will for sure create a new issue about it.
